pkg/cloud/baremetal/actuators/machine/actuator.go

```diff
@@ -584,3 +593,163 @@ func (a *Actuator) nodeAddresses(host *bmh.BareMetalHost) ([]corev1.NodeAddress,
+	if baremetalhost.Annotations == nil {
+		baremetalhost.Annotations = make(map[string]string)
+	}
+
+	baremetalhost.Annotations[rebootAnnotation] = ""
+
+	err := a.client.Update(ctx, baremetalhost)
+	if err != nil {
+		log.Printf("failed to add reboot annotation to %s: %s", baremetalhost.Name, err.Error())
+	}
+
+	return err
+}
+
+//requestPowerOn adds reboot annotation on baremetalhost which signal BMO to power off the machine
+func (a *Actuator) requestPowerOn(ctx context.Context, baremetalhost *bmh.BareMetalHost) error {
+	if baremetalhost.Annotations == nil {
+		baremetalhost.Annotations = make(map[string]string)
+	}
+
+	if _, rebootPending := baremetalhost.Annotations[rebootAnnotation]; !rebootPending {
+		return &clustererror.RequeueAfterError{}
+	}
+
+	delete(baremetalhost.Annotations, rebootAnnotation)
+
+	err := a.client.Update(ctx, baremetalhost)
+	if err != nil {
+		log.Printf("failed to remove reboot annotation from %s: %s", baremetalhost.Name, err.Error())
+	}
+
+	return err
+}
+
+// deleteMachineNode deletes the node that mapped to specified machine
+func (a *Actuator) deleteNode(ctx context.Context, node *corev1.Node) error {
+	err := a.client.Delete(ctx, node)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return &clustererror.RequeueAfterError{}
+		}
+		log.Printf("Failed to delete node %s: %s", node.Name, err.Error())
+	}
+	return err
+}
+
+// getNodeByMachine returns the node object referenced by machine
+func (a *Actuator) getNodeByMachine(ctx context.Context, machine *machinev1.Machine) (*corev1.Node, error) {
+	if machine.Status.NodeRef == nil {
+		return nil, errors.NewNotFound(corev1.Resource("ObjectReference"), machine.Name)
+	}
+
+	node := &corev1.Node{}
+	key := client.ObjectKey{
+		Name:      machine.Status.NodeRef.Name,
+		Namespace: machine.Status.NodeRef.Namespace,
+	}
+
+	if err := a.client.Get(ctx, key, node); err != nil {
+		return nil, err
+	}
+	return node, nil
+}
+
+//remediateIfNeeded will try to remediate unhealthy nodes (annotated by MHC)
+func (a *Actuator) remediateIfNeeded(ctx context.Context, machine *machinev1.Machine, baremetalhost *bmh.BareMetalHost) error {
+	if len(machine.Annotations) == 0 {
+		return nil
+	}
+
+	if _, needsRemediation := machine.Annotations[externalRemediationAnnotation]; !needsRemediation {
+		return nil
+	}
+
+	if _, remediationInProgress := machine.Annotations[remediationInProgressAnnotation]; !remediationInProgress {
+		if !hasRebootAnnotation(baremetalhost) {
+			log.Printf("Found an unhealthy machine, requesting power off. Machine name: %s", machine.Name)
+			return a.requestPowerOff(ctx, baremetalhost)
+		}
+
+		if baremetalhost.Status.PoweredOn {
+			return nil
+		}
+
+		//we need this annotation to differentiate between unhealthy machine that
+		//needs remediation to unhealthy machine that just got remediated
+		return a.addRemediationInProgressAnnotation(ctx, machine)
```

Comment:
I think this workflow is starting to make sense to me. Does this annotation really mean that "We requested power off, and then observed that power transitioned from on to off"? I could argue that remediation is "in progress" as soon as the power-off is requested, so if I'm understanding correctly, I think this name could throw people off (and it did throw me off). Maybe it would be easier to understand this workflow if this annotation was named something more directly representative of what it's measuring. Maybe `HostPoweredOffAnnotation`?

Reply:
If it would be `HostPoweredOffAnnotation` we could argue that it's not always true, since we remove this annotation after the host was powered on.
The main motivation for this annotation is described in the comments (two lines before this one). 

I think it's a good name but I'm probably biased and my head was in this code for too long, and I'm blind to such issues.
@beekhof - what do you think?

Reply:
I wonder why, if we can tell that we need to add this annotation, we can't use the same conditions to tell that we have a machine in the middle of remediation elsewhere. Why do we need to store a separate piece of state information?

Reply:
We can tell that we need to add this annotation while we're on specific state. In later state, where we encounter powered-on host, with unhealthy annotation - we can't tell if this is a result of successful remediation or a real unhealthy machine that needs a remediation.
In addition, we can't remove the unhealthy annotation at this point, as MHC might annotate it again and again (since the Machine is still not healthy)